rmw_fastrtps_cpp/src/subscription.cpp

```diff
@@ -106,86 +123,143 @@ create_subscription(
-  subscriberParam.topic.topicDataType = type_name;
-  subscriberParam.topic.topicName = _create_topic_name(qos_policies, ros_topic_prefix, topic_name);
+  info->type_support_ = new (std::nothrow) MessageTypeSupport_cpp(callbacks);
+  if (!info->type_support_) {
+    RMW_SET_ERROR_MSG("create_subscription() failed to allocate MessageTypeSupport");
+    return nullptr;
+  }
+  auto cleanup_type_support = rcpputils::make_scope_exit(
+    [info]() {
+      delete info->type_support_;
+    });
 
-  if (!get_datareader_qos(*qos_policies, subscriberParam)) {
+  /////
+  // Register the Type in the participant
+  // When a type is registered in a participant, it is converted to a shared_ptr, so it is
+  // dangerous to keep using it. Thus we use a new TypeSupport created only to register it.
+  ReturnCode_t ret = domainParticipant->register_type(
+    eprosima::fastdds::dds::TypeSupport(
+      new (std::nothrow) MessageTypeSupport_cpp(callbacks)));
+  // Register could fail if there is already a type with that name in participant,
+  // so not only OK retcode is possible
+  if (ret != ReturnCode_t::RETCODE_OK && ret != ReturnCode_t::RETCODE_PRECONDITION_NOT_MET) {
     return nullptr;
   }
 
+  /////
+  // Create Listener
+  info->listener_ = nullptr;
   if (create_subscription_listener) {
     info->listener_ = new (std::nothrow) SubListener(info);
-    if (!info->listener_) {
-      RMW_SET_ERROR_MSG("create_subscriber() could not create subscriber listener");
-      return nullptr;
-    }
   }
 
-  info->subscriber_ = Domain::createSubscriber(participant, subscriberParam, info->listener_);
-  if (!info->subscriber_) {
-    RMW_SET_ERROR_MSG("create_subscriber() could not create subscriber");
+  if (!info->listener_) {
+    RMW_SET_ERROR_MSG("create_subscription() could not create subscription listener");
     return nullptr;
   }
-  auto cleanup_subscription = rcpputils::make_scope_exit(
+
+  auto cleanup_listener = rcpputils::make_scope_exit(
     [info]() {
-      if (!Domain::removeSubscriber(info->subscriber_)) {
-        RMW_SAFE_FWRITE_TO_STDERR(
-          "Failed to remove subscriber after '"
-          RCUTILS_STRINGIFY(__function__) "' failed.\n");
-      }
+      delete info->listener_;
+    });
+
+  /////
+  // Create Topic
+
+  // Create Topic name
+  auto topic_name_mangled =
+    _create_topic_name(qos_policies, ros_topic_prefix, topic_name).to_string();
+
+  // Use Topic Qos Default
+  eprosima::fastdds::dds::TopicQos topicQos = domainParticipant->get_default_topic_qos();
+
+  if (!get_topic_qos(*qos_policies, topicQos)) {
+    return nullptr;
+  }
+
+  // General function to create or get an already existing topic
+  eprosima::fastdds::dds::TopicDescription * des_topic =
+    rmw_fastrtps_shared_cpp::create_topic_rmw(
+    participant_info,
+    topic_name_mangled,
+    type_name,
+    topicQos);
+
+  if (des_topic == nullptr) {
+    RMW_SET_ERROR_MSG("create_subscription() failed to create topic");
+    return nullptr;
+  }
+
```

Comment:
There is no `dynamic_cast` to `Topic` Entity just in case there are future specialization of base class `TopicDescription`, as it is done in `publisher.cpp`.

Reply:
As `create_datareader` receives a `TopicDescription*`, the cast is only necessary for writers